File: utils/languageDetection.ts
/**
 * Language Detection Utility
 * 
 * Provides client-side language detection for user input text
 * using character patterns and common word detection.
 */

export type DetectedLanguage = {
  code: string;
  confidence: number;
  name: string;
};

// Language patterns for detection
const LANGUAGE_PATTERNS = {
  en: {
    name: 'English',
    chars: /[a-zA-Z]/,
    commonWords: ['the', 'is', 'are', 'and', 'of', 'to', 'in', 'for', 'with', 'on']
  },
  hi: {
    name: 'Hindi',
    chars: /[\u0900-\u097F]/,
    commonWords: ['है', 'हैं', 'का', 'के', 'की', 'में', 'से', 'को', 'पर', 'ने']
  },
  ta: {
    name: 'Tamil',
    chars: /[\u0B80-\u0BFF]/,
    commonWords: ['இது', 'அது', 'இந்த', 'அந்த', 'என்ன', 'எப்படி', 'எங்கே', 'யார்']
  },
  te: {
    name: 'Telugu',
    chars: /[\u0C00-\u0C7F]/,
    commonWords: ['ఇది', 'అది', 'ఈ', 'ఆ', 'ఏమి', 'ఎలా', 'ఎక్కడ', 'ఎవరు']
  },
  bn: {
    name: 'Bengali',
    chars: /[\u0980-\u09FF]/,
    commonWords: ['এটি', 'ওটা', 'এই', 'ওই', 'কি', 'কেমন', 'কোথায়', 'কে']
  },
  mr: {
    name: 'Marathi',
    chars: /[\u0900-\u097F]/,
    commonWords: ['हे', 'ते', 'या', 'त्या', 'काय', 'कसे', 'कुठे', 'कोण']
  },
  gu: {
    name: 'Gujarati',
    chars: /[\u0A80-\u0AFF]/,
    commonWords: ['આ', 'તે', 'આ', 'તે', 'શું', 'કેમ', 'ક્યાં', 'કોણ']
  },
  kn: {
    name: 'Kannada',
    chars: /[\u0C80-\u0CFF]/,
    commonWords: ['ಇದು', 'ಅದು', 'ಈ', 'ಆ', 'ಏನು', 'ಹೇಗೆ', 'ಎಲ್ಲಿ', 'ಯಾರು']
  },
  ml: {
    name: 'Malayalam',
    chars: /[\u0D00-\u0D7F]/,
    commonWords: ['ഇത്', 'അത്', 'ഈ', 'ആ', 'എന്ത്', 'എങ്ങനെ', 'എവിടെ', 'ആര്']
  },
  pa: {
    name: 'Punjabi',
    chars: /[\u0A00-\u0A7F]/,
    commonWords: ['ਇਹ', 'ਉਹ', 'ਇਸ', 'ਉਸ', 'ਕੀ', 'ਕਿਵੇਂ', 'ਕਿੱਥੇ', 'ਕੌਣ']
  },
  es: {
    name: 'Spanish',
    chars: /[a-zA-ZáéíóúñÁÉÍÓÚÑ]/,
    commonWords: ['el', 'la', 'de', 'que', 'y', 'es', 'en', 'por', 'un', 'para']
  },
  fr: {
    name: 'French',
    chars: /[a-zA-ZàâäæçéèêëïîôùûüÿœÀÂÄÆÇÉÈÊËÏÎÔÙÛÜŸŒ]/,
    commonWords: ['le', 'la', 'de', 'et', 'un', 'une', 'dans', 'pour', 'que', 'qui']
  },
  de: {
    name: 'German',
    chars: /[a-zA-ZäöüßÄÖÜ]/,
    commonWords: ['der', 'die', 'das', 'und', 'in', 'ist', 'von', 'den', 'zu', 'mit']
  },
  pt: {
    name: 'Portuguese',
    chars: /[a-zA-ZáàâãéêíóôõúçÁÀÂÃÉÊÍÓÔÕÚÇ]/,
    commonWords: ['o', 'a', 'de', 'que', 'e', 'do', 'da', 'em', 'um', 'para']
  },
  ja: {
    name: 'Japanese',
    chars: /[\u3040-\u309F\u30A0-\u30FF\u4E00-\u9FAF]/,
    commonWords: ['の', 'に', 'は', 'を', 'た', 'が', 'で', 'て', 'と', 'し']
  },
  'zh-CN': {
    name: 'Chinese',
    chars: /[\u4E00-\u9FFF]/,
    commonWords: ['的', '是', '在', '了', '和', '有', '我', '你', '他', '她']
  }
};

/**
 * Detect the language of input text
 */
export function detectLanguage(text: string): DetectedLanguage {
  if (!text || text.trim().length === 0) {
    return { code: 'en', confidence: 0, name: 'English' };
  }

  const normalizedText = text.toLowerCase().trim();
  const scores: Record<string, number> = {};

  // Score based on character patterns
  for (const [langCode, pattern] of Object.entries(LANGUAGE_PATTERNS)) {
    let score = 0;
    
    // Count matching characters
    const chars = normalizedText.match(pattern.chars);
    if (chars) {
      score += (chars.length / normalizedText.length) * 50;
    }

    // Check for common words
    const words = normalizedText.split(/\s+/);
    for (const word of words) {
      if (pattern.commonWords.some(commonWord => 
        word.includes(commonWord) || commonWord.includes(word)
      )) {
        score += 10;
      }
    }

    scores[langCode] = score;
  }

  // Find language with highest score
  let maxScore = 0;
  let detectedLang = 'en';

  for (const [langCode, score] of Object.entries(scores)) {
    if (score > maxScore) {
      maxScore = score;
      detectedLang = langCode;
    }
  }

  // Calculate confidence (0-1 scale)
  const confidence = Math.min(maxScore / 100, 1);

  return {
    code: detectedLang,
    confidence,
    name: LANGUAGE_PATTERNS[detectedLang as keyof typeof LANGUAGE_PATTERNS].name
  };
}

/**
 * Check if text contains mixed languages
 */
export function detectMixedLanguages(text: string): DetectedLanguage[] {
  if (!text || text.trim().length === 0) {
    return [];
  }

  const detected: DetectedLanguage[] = [];
  const scores: Record<string, number> = {};

  // Score all languages
  for (const [langCode, pattern] of Object.entries(LANGUAGE_PATTERNS)) {
    const chars = text.match(pattern.chars);
    if (chars && chars.length > 0) {
      const score = (chars.length / text.length) * 100;
      if (score > 10) { // Only include if >10% presence
        scores[langCode] = score;
      }
    }
  }

  // Convert to array and sort by score
  for (const [langCode, score] of Object.entries(scores)) {
    detected.push({
      code: langCode,
      confidence: score / 100,
      name: LANGUAGE_PATTERNS[langCode as keyof typeof LANGUAGE_PATTERNS].name
    });
  }

  return detected.sort((a, b) => b.confidence - a.confidence);
}

/**
 * Get user's preferred language from browser settings
 */
export function getBrowserLanguage(): string {
  if (typeof window === 'undefined' || !navigator.languages) {
    return 'en';
  }

  // Check browser language settings
  for (const lang of navigator.languages) {
    const langCode = lang.split('-')[0];
    if (LANGUAGE_PATTERNS[langCode as keyof typeof LANGUAGE_PATTERNS]) {
      return langCode;
    }
  }

  return 'en';
}

/**
 * Validate if a language code is supported
 */
export function isSupportedLanguage(code: string): boolean {
  return code in LANGUAGE_PATTERNS;
}

/**
 * Get language name from code
 */
export function getLanguageName(code: string): string {
  const pattern = LANGUAGE_PATTERNS[code as keyof typeof LANGUAGE_PATTERNS];
  return pattern ? pattern.name : 'Unknown';
}

File: backend/routes/language_detection.py
"""
Language Detection API Route

Provides language detection for user input text to enhance multilingual support.
"""

import logging
from typing import Any

from fastapi import APIRouter, Request
from pydantic import BaseModel, Field

from backend.middleware.rate_limiter import CHAT_RATE_LIMIT, limiter

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/language", tags=["language"])


class LanguageDetectionRequest(BaseModel):
    """Request model for language detection."""
    text: str = Field(..., min_length=1, max_length=1000, description="Text to detect language from")


class LanguageDetectionResponse(BaseModel):
    """Response model for language detection."""
    detected_language: str = Field(..., description="Detected language code (e.g., 'en', 'hi', 'es')")
    confidence: float = Field(..., ge=0, le=1, description="Confidence score (0-1)")
    supported: bool = Field(..., description="Whether language is supported by KIAAN")
    language_name: str = Field(..., description="Full name of detected language")


# Language patterns for simple detection
LANGUAGE_PATTERNS = {
    'en': {
        'name': 'English',
        'chars': r'[a-zA-Z]',
        'sample': 'the is are and'
    },
    'hi': {
        'name': 'Hindi',
        'chars': r'[\u0900-\u097F]',
        'sample': 'है हैं का'
    },
    'ta': {
        'name': 'Tamil',
        'chars': r'[\u0B80-\u0BFF]',
        'sample': 'இது அது'
    },
    'te': {
        'name': 'Telugu',
        'chars': r'[\u0C00-\u0C7F]',
        'sample': 'ఇది అది'
    },
    'bn': {
        'name': 'Bengali',
        'chars': r'[\u0980-\u09FF]',
        'sample': 'এটি ওটা'
    },
    'mr': {
        'name': 'Marathi',
        'chars': r'[\u0900-\u097F]',
        'sample': 'हे ते'
    },
    'gu': {
        'name': 'Gujarati',
        'chars': r'[\u0A80-\u0AFF]',
        'sample': 'આ તે'
    },
    'kn': {
        'name': 'Kannada',
        'chars': r'[\u0C80-\u0CFF]',
        'sample': 'ಇದು ಅದು'
    },
    'ml': {
        'name': 'Malayalam',
        'chars': r'[\u0D00-\u0D7F]',
        'sample': 'ഇത് അത്'
    },
    'pa': {
        'name': 'Punjabi',
        'chars': r'[\u0A00-\u0A7F]',
        'sample': 'ਇਹ ਉਹ'
    },
    'es': {
        'name': 'Spanish',
        'chars': r'[a-zA-ZáéíóúñÁÉÍÓÚÑ]',
        'sample': 'el la de'
    },
    'fr': {
        'name': 'French',
        'chars': r'[a-zA-ZàâäæçéèêëïîôùûüÿœÀÂÄÆÇÉÈÊËÏÎÔÙÛÜŸŒ]',
        'sample': 'le la de'
    },
    'de': {
        'name': 'German',
        'chars': r'[a-zA-ZäöüßÄÖÜ]',
        'sample': 'der die das'
    },
    'pt': {
        'name': 'Portuguese',
        'chars': r'[a-zA-ZáàâãéêíóôõúçÁÀÂÃÉÊÍÓÔÕÚÇ]',
        'sample': 'o a de'
    },
    'ja': {
        'name': 'Japanese',
        'chars': r'[\u3040-\u309F\u30A0-\u30FF\u4E00-\u9FAF]',
        'sample': 'の に は'
    },
    'zh-CN': {
        'name': 'Chinese',
        'chars': r'[\u4E00-\u9FFF]',
        'sample': '的 是 在'
    }
}


def simple_detect_language(text: str) -> tuple[str, float]:
    """
    Simple language detection based on character patterns.
    
    Args:
        text: Text to detect language from
        
    Returns:
        Tuple of (language_code, confidence_score)
    """
    import re
    
    if not text or not text.strip():
        return ('en', 0.0)
    
    scores = {}
    
    # Score based on character presence
    for lang_code, pattern_info in LANGUAGE_PATTERNS.items():
        pattern = pattern_info['chars']
        matches = re.findall(pattern, text)
        if matches:
            score = len(matches) / len(text)
            scores[lang_code] = score
    
    if not scores:
        return ('en', 0.5)  # Default to English with medium confidence
    
    # Get language with highest score
    detected_lang = max(scores, key=scores.get)
    confidence = min(scores[detected_lang] * 2, 1.0)  # Scale to 0-1
    
    return (detected_lang, confidence)


@router.post("/detect", response_model=LanguageDetectionResponse)
@limiter.limit(CHAT_RATE_LIMIT)
async def detect_language(request: Request, data: LanguageDetectionRequest) -> dict[str, Any]:
    """
    Detect the language of input text.
    
    This endpoint provides simple language detection based on character patterns.
    For production use, consider integrating with Google Cloud Translation API
    or similar services for more accurate detection.
    
    Args:
        request: FastAPI request object
        data: Language detection request data
        
    Returns:
        Language detection result
    """
    try:
        text = data.text.strip()
        
        if not text:
            return {
                "detected_language": "en",
                "confidence": 0.0,
                "supported": True,
                "language_name": "English"
            }
        
        # Detect language
        detected_lang, confidence = simple_detect_language(text)
        
        # Check if supported
        supported = detected_lang in LANGUAGE_PATTERNS
        language_name = LANGUAGE_PATTERNS.get(detected_lang, {}).get('name', 'Unknown')
        
        logger.info(f"Language detected: {detected_lang} (confidence: {confidence:.2f})")
        
        return {
            "detected_language": detected_lang,
            "confidence": confidence,
            "supported": supported,
            "language_name": language_name
        }
        
    except Exception as e:
        logger.error(f"Language detection error: {e}")
        # Return default on error
        return {
            "detected_language": "en",
            "confidence": 0.5,
            "supported": True,
            "language_name": "English"
        }


@router.get("/supported")
async def get_supported_languages() -> dict[str, Any]:
    """
    Get list of supported languages for KIAAN.
    
    Returns:
        Dictionary of supported languages with codes and names
    """
    return {
        "supported_languages": {
            code: info['name']
            for code, info in LANGUAGE_PATTERNS.items()
        },
        "count": len(LANGUAGE_PATTERNS)
    }
